packages/jellyfish-api-core/__tests__/category/loan/withdrawFromVault.test.ts

```diff
@@ -429,3 +429,367 @@ describe('Loan', () => {
+    expect(vaultAfter.collateralAmounts).toStrictEqual(['5000.00000000@DUSD', '1.00000000@BTC'])
+    expect(vaultAfter.collateralValue).toStrictEqual(new BigNumber(10000))
+  })
+
+  it('should withdrawFromVault with 50% DUSD of minimum required collateral', async () => {
+    // add btc collateral, new total collateral = 20000 USD
+    await alice.rpc.loan.depositToVault({
+      vaultId: bobVaultId, from: aliceAddr, amount: '1@BTC' // collateral value = 1 x 10000 x 0.5 = 5000 USD
+    })
+    await alice.generate(1)
+
+    // must wait until block count reaches fort canning road height
+    const blockCount = await bob.container.getBlockCount()
+    await bob.generate(fortCanningRoadHeight - blockCount)
+    await tGroup.waitForSync()
+
+    const destinationAddress = await alice.generateAddress()
+    const accountBalancesBefore = await alice.rpc.account.getAccount(destinationAddress)
+    expect(accountBalancesBefore.length).toStrictEqual(0)
+
+    const tslaLoanAmount = 2500 // loan amount = 5000 USD
+    await bob.rpc.loan.takeLoan({
+      vaultId: bobVaultId,
+      amounts: `${tslaLoanAmount}@TSLA`
+    })
+    // remove dfi collateral, new total collateral = 10000 USD
+    await bob.rpc.loan.withdrawFromVault({
+      vaultId: bobVaultId, to: destinationAddress, amount: '5000@DFI'
+    })
+    await bob.generate(1)
+    await tGroup.waitForSync()
+
+    const accountBalancesAfter = await alice.rpc.account.getAccount(destinationAddress)
+    expect(accountBalancesAfter.length).toStrictEqual(1)
+    expect(accountBalancesAfter[0]).toStrictEqual('5000.00000000@DFI')
+
+    const vaultAfter = await bob.rpc.loan.getVault(bobVaultId) as VaultActive
+    expect(vaultAfter.collateralAmounts).toStrictEqual(['5000.00000000@DUSD', '2.00000000@BTC'])
+    expect(vaultAfter.collateralValue).toStrictEqual(new BigNumber(15000))
+  })
+
+  it('should withdrawFromVault with 25% DFI + 25% DUSD collateral', async () => {
+    // must wait until block count reaches fort canning road height
+    const blockCount = await bob.container.getBlockCount()
+    await bob.generate(fortCanningRoadHeight - blockCount)
+    await tGroup.waitForSync()
+
+    const destinationAddress = await alice.generateAddress()
+    const accountBalancesBefore = await alice.rpc.account.getAccount(destinationAddress)
+    expect(accountBalancesBefore.length).toStrictEqual(0)
+
+    const tslaLoanAmount = 2500 // loan amount = 5000 USD
+    await bob.rpc.loan.takeLoan({
+      vaultId: bobVaultId,
+      amounts: `${tslaLoanAmount}@TSLA`
+    })
+    // remove dfi collateral, new total collateral = 12500 USD
+    await bob.rpc.loan.withdrawFromVault({
+      vaultId: bobVaultId, to: destinationAddress, amount: '2500@DFI'
+    })
+    // remove dusd collateral, new total collateral = 10000 USD
+    await bob.rpc.loan.withdrawFromVault({
+      vaultId: bobVaultId, to: destinationAddress, amount: '2500@DUSD'
+    })
```

Comment:
between these two `withdrawFromVault` actions, there's no block generated. Hence the changes are not applied to the blockchain until a block is minted. 
This means that for this regtest setup, both txs are going to be in the same block.
Always try to generate a block after some action that change blockchain's state, unless you specifically require it to be in the same block.
This is what i mean "fix 1" in https://github.com/DeFiCh/jellyfish/pull/1205#discussion_r841166931  😅

Reply:
Here both actions required to be in the same block, otherwise we will not be able to test withdrawing exactly 25% DFI and 25% DUSD since interest adds after taking loan.

Reply:
If it mandatory to generate a block after a withdraw, I can move one withdraw above the take loan.